tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ArcHybridTransitionSystem.scala

```diff
@@ -90,49 +90,87 @@ case class ArcHybridTransitionSystem(
-      PreviousLinkGrandgretelRef, StackGretelsRef(0), StackGretelsRef(1), StackLeftGretelsRef(0),
-      StackRightGretelsRef(0), BufferGretelsRef(0))),
-    new OfflineTokenFeature(StackRef(0)),
-    new OfflineTokenFeature(StackRef(1)),
-    new OfflineTokenFeature(StackRef(2)),
-    new OfflineTokenFeature(BufferRef(0)),
-    new OfflineTokenFeature(BufferRef(1)),
-    new OfflineTokenFeature(PreviousLinkCrumbRef),
-    new OfflineTokenFeature(PreviousLinkGretelRef),
-    new OfflineTokenFeature(PreviousLinkCrumbGretelRef),
-    new OfflineTokenFeature(PreviousLinkGrandgretelRef),
-    new OfflineTokenFeature(StackGretelsRef(0)),
-    new OfflineTokenFeature(StackLeftGretelsRef(0)),
-    new OfflineTokenFeature(StackRightGretelsRef(0)),
-    new OfflineTokenFeature(StackGretelsRef(1)),
-    new OfflineTokenFeature(BufferGretelsRef(0)),
-    new OfflineTokenFeature(LastRef),
-    new OfflineTokenFeature(FirstRef)
-  ))
-
-  val transitionFeature = FeatureUnion(List(
-    new TokenCardinalityFeature(Seq(StackRef(0), StackRef(1), StackRef(2), BufferRef(0),
-      BufferRef(1), StackGretelsRef(0), StackGretelsRef(1), StackLeftGretelsRef(0),
-      StackRightGretelsRef(0), BufferGretelsRef(0))),
-    new OfflineTokenFeature(StackRef(0)),
-    new OfflineTokenFeature(StackRef(1)),
-    new OfflineTokenFeature(StackRef(2)),
-    new OfflineTokenFeature(BufferRef(0)),
-    new OfflineTokenFeature(BufferRef(1)),
-    new OfflineTokenFeature(StackChildrenRef(0)),
-    new OfflineTokenFeature(StackChildrenRef(1)),
-    new OfflineTokenFeature(BufferChildrenRef(0)),
-    new OfflineTokenFeature(StackGretelsRef(0)),
-    new OfflineTokenFeature(StackLeftGretelsRef(0)),
-    new OfflineTokenFeature(StackRightGretelsRef(0)),
-    new OfflineTokenFeature(StackGretelsRef(1)),
-    new OfflineTokenFeature(BufferGretelsRef(0)),
-    new OfflineTokenFeature(LastRef),
-    new OfflineTokenFeature(FirstRef)
-  ))
+  private val previousLinkFeatures: Seq[StateFeature] = (for {
+    stateRef <- Seq(PreviousLinkCrumbRef, PreviousLinkGretelRef)
+    neighbors <- Seq(
+      Seq(TokenChildren), Seq(TokenParents),
+      Seq(TokenChild(0)), Seq(TokenChild(1)), Seq(TokenChild(2)), Seq(TokenChild(3)),
+      Seq(TokenParent(0)), Seq(TokenParent(1)), Seq(TokenParent(2)), Seq(TokenParent(3))
+    )
+  } yield {
+    Seq(
+      new TokenLinkFeature(stateRef, TransitiveRef(stateRef, neighbors))
+    )
+  }).flatten
+
+  private val linkFeatures: Seq[StateFeature] = (for {
+    stateRef <- Seq(StackRef(0), StackRef(1), BufferRef(0))
+    neighbors <- Seq(
+      Seq(TokenChildren), Seq(TokenParents),
+      Seq(TokenChild(0)), Seq(TokenChild(1)), Seq(TokenChild(2)), Seq(TokenChild(3)),
+      Seq(TokenParent(0)), Seq(TokenParent(1)), Seq(TokenParent(2)), Seq(TokenParent(3))
+    )
+  } yield {
+    Seq(
+      new TokenLinkFeature(stateRef, TransitiveRef(stateRef, neighbors))
+    )
+  }).flatten
```

Comment:
Both of these for loops create a `Seq[Seq[T]]`, and then call `flatten`, but the inner `Seq` always has only one element. You could skip one `Seq` and the call to `flatten`.